### ISSUE TITLE
Update not working until chaos restart

### ISSUE THREAD

**Issue by Yiyiyimu** [chaos-mesh/chaos-mesh#1259]
## Bug Report

**What version of Kubernetes are you using?**
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:58:59Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:43:34Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}

**What did you do?**

Install chaos-mesh via `curl -sSL https://mirrors.chaos-mesh.org/v1.0.2/install.sh | bash`

For immediate chaos applied on the dashboard, I found I could update the chaos, but it would not take effect (What I test is network delay for NetworkChaos). After some test I found it would only update to the new parameters when the chaos restart (either pause/unpause, or wait for it to start another cron cycle). 
Considering it might be hard to change the chaos status at once when running, I guess we could either
1. restart the chaos to make it take effect
2. inform users it would not take effect at once

**What did you expect to see?**
Chaos take effect at once

**Comment by Yiyiyimu** [comment]
Hi @cwen0 I think it is not only the problem for UI, it's actually another problem of two phase controller. 

When I update the chaos, the desired procedure should be "recover - start" to cancel the original chaos and apply the new one. However in the current implementation, the reconciler would go into phase Waiting and do nothing, so we need to wait until the next start to make it effective.

The question is, if spec is got changed, how could we tell it changed? I mean it seems not appropriate to save a chaos configuration somewhere and compare the new one with it. Besides we do hope it could be stateless. Do you have some suggestions?

(And indeed, the whole scheduler would get more and more complicated :(

**Comment by cwen0** [comment]
> Hi @cwen0 I think it is not only the problem for UI, it's actually another problem of two phase controller.
> 
> When I update the chaos, the desired procedure should be "recover - start" to cancel the original chaos and apply the new one. However in the current implementation, the reconciler would go into phase Waiting and do nothing, so we need to wait until the next start to make it effective.
> 
> The question is, if spec is got changed, how could we tell it changed? I mean it seems not appropriate to save a chaos configuration somewhere and compare the new one with it. Besides we do hope it could be stateless. Do you have some suggestions?
> 
> (And indeed, the whole scheduler would get more and more complicated :(

Yes, this is not only the problem for UI, we can inform users it would not take effect at once on UI first.  Maybe this function  [WithEventFilter](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/builder#Builder.WithEventFilter) can help us to filter and compare the old and new configureatiron.

**Comment by Yiyiyimu** [comment]
`WithEventFilter` seems cool! Thanks for the hint! I'll take care of it later

**Comment by YangKeao** [comment]
@Yiyiyimu If you need some sample codes about `EventFilter`, you can find it in #1221 ([these lines](https://github.com/chaos-mesh/chaos-mesh/pull/1221/files#diff-40c2ca106c38ecb998a3158bb60025dd3732c09c5ab4b4d634ddd39953932790R148-R171)).

Using multiple controllers seem to be a good way to split the logic.

(Actually I am considering an event driven controller design, by defining and registering more detailed events and stepping from event loop to callback and finally in `async/await` synchronized pattern (though in go we don't need `async/await`, blocking `goroutine` is enough :smile: ). It could be helpful for reimplementing `pod*chaos` with `reconciler` and the implementation of workflow (especially for the combination and reaction about multiple resources). But it's still an imagination, and still fuzzy in my mind :cry: .)

**Comment by Yiyiyimu** [comment]
@YangKeao Thanks for the sample codes! 

>(Actually I am considering an event driven controller design, by defining and registering more detailed events and stepping from event loop to callback and finally in async/await synchronized pattern (though in go we don't need async/await, blocking goroutine is enough 😄 ). 

I can see what you mean but indeed it's hard for me to figure out the detailed implementations for now🙈. Let's talk about it sometime later

**Comment by TobiColada** [comment]
👍


Von meinem iPhone gesendet
